spec/std/xml/xml_spec.cr

```diff
@@ -168,45 +168,205 @@ describe XML do
+      <?xml version="1.0" encoding="UTF-8"?>
+      <feed xmlns="http://www.w3.org/2005/Atom" xmlns:openSearch="http://a9.com/-/spec/opensearchrss/1.0/">
+        <item></item>
+      </feed>
+      XML
+      )
+
+      doc.root.not_nil!.first_element_child.not_nil!.namespace_definitions.should be_empty
+    end
   end
 
-  it "returns empty array if no namespaces scopes exists" do
-    doc = XML.parse(<<-XML
+  describe "#namespace_scopes" do
+    it "gets root namespaces scopes" do
+      doc = XML.parse(<<-XML
+      <?xml version="1.0" encoding="UTF-8"?>
+      <feed xmlns="http://www.w3.org/2005/Atom" xmlns:openSearch="http://a9.com/-/spec/opensearchrss/1.0/">
+      </feed>
+      XML
+      )
+      namespaces = doc.root.not_nil!.namespace_scopes
+
+      namespaces.size.should eq(2)
+      namespaces[0].href.should eq("http://www.w3.org/2005/Atom")
+      namespaces[0].prefix.should be_nil
+      namespaces[1].href.should eq("http://a9.com/-/spec/opensearchrss/1.0/")
+      namespaces[1].prefix.should eq("openSearch")
+    end
+
+    it "returns empty array if no namespaces scopes exists" do
+      doc = XML.parse(<<-XML
       <?xml version='1.0' encoding='UTF-8'?>
       <name>John</name>
       XML
-    )
-    namespaces = doc.root.not_nil!.namespace_scopes
+      )
+      namespaces = doc.root.not_nil!.namespace_scopes
 
-    namespaces.size.should eq(0)
+      namespaces.size.should eq(0)
+    end
+
+    it "includes parent namespaces" do
+      doc = XML.parse(<<-XML
+      <?xml version="1.0" encoding="UTF-8"?>
+      <feed xmlns="http://www.w3.org/2005/Atom" xmlns:openSearch="http://a9.com/-/spec/opensearchrss/1.0/">
+        <item xmlns:c="http://c"></item>
+      </feed>
+      XML
+      )
+      namespaces = doc.root.not_nil!.first_element_child.not_nil!.namespace_scopes
+
+      namespaces.size.should eq(3)
+      namespaces[0].href.should eq("http://c")
+      namespaces[0].prefix.should eq "c"
+      namespaces[1].href.should eq("http://www.w3.org/2005/Atom")
+      namespaces[1].prefix.should be_nil
+      namespaces[2].href.should eq("http://a9.com/-/spec/opensearchrss/1.0/")
+      namespaces[2].prefix.should eq("openSearch")
+    end
   end
 
-  it "gets root namespaces as hash" do
-    doc = XML.parse(<<-XML
+  describe "#namespaces" do
+    it "gets root namespaces as hash" do
+      doc = XML.parse(<<-XML
       <?xml version="1.0" encoding="UTF-8"?>
       <feed xmlns="http://www.w3.org/2005/Atom" xmlns:openSearch="http://a9.com/-/spec/opensearchrss/1.0/">
       </feed>
       XML
-    )
-    namespaces = doc.root.not_nil!.namespaces
-    namespaces.should eq({
-      "xmlns"          => "http://www.w3.org/2005/Atom",
-      "xmlns:openSearch": "http://a9.com/-/spec/opensearchrss/1.0/",
-    })
+      )
+      namespaces = doc.root.not_nil!.namespaces
+      namespaces.should eq({
+        "xmlns"          => "http://www.w3.org/2005/Atom",
+        "xmlns:openSearch": "http://a9.com/-/spec/opensearchrss/1.0/",
```

Comment:
How does this hyprid literal even compile? 🕵️

Reply:
Oh interesting.  Apparently it was like that before too :thinking:.

Reply:
https://github.com/crystal-lang/crystal/issues/11074